core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -101,20 +100,34 @@ private String fileClass() {
 
   protected ManifestReader(
       InputFile file,
+      int specId,
       Map<Integer, PartitionSpec> specsById,
       InheritableMetadata inheritableMetadata,
       FileType content) {
     this.file = file;
     this.inheritableMetadata = inheritableMetadata;
     this.content = content;
 
+    if (specsById != null) {
+      this.spec = specsById.get(specId);
+      Preconditions.checkArgument(
+          spec != null, "Could not find PartitionSpec for specId: %s", specId);
+    } else {
+      this.spec = readPartitionSpec(file);
+    }
+
+    this.fileSchema = new Schema(DataFile.getType(spec.partitionType()).fields());
+  }
+
+  private PartitionSpec readPartitionSpec(InputFile inputFile) {
```

Comment:
Can this be `static`?

Reply:
This should still pass in `specsById`. There is no reason to change the behavior of loading. If the spec is already loaded and defined, it should be used rather than parsed.

Reply:
Updated to `static` and changed to use `specsById` if it is present.